File: Solutions/AddBinary.cpp
// Leetcode #67: https://leetcode.com/problems/add-binary/
// Source: https://leetcode.com/problems/add-binary/discuss/1679423/Well-Detailed-Explaination-Java-C%2B%2B-Python-oror-Easy-for-mind-to-Accept-it

// Solution 1 | O(n) time, O(1) space
class Solution {
public:
    string addBinary(string a, string b) {
        // Initialize helper variables
        int i = a.length() - 1;
        int j = b.length() - 1;
        int carry = 0;
        string result;
        
        // While the strings have valid indices or carry > 0
        while (i >= 0 || j >= 0 || carry > 0) {
            // Initialize temp with value of carry and add value of a[i] and b[j] if they exist. 
            //Subtract '0' gets the integer value.
            int temp = carry;
            if (i >= 0) temp += a[i--] - '0';
            if (j >= 0) temp += b[j--] - '0';
        
            // The binary digit to add is temp % 2.
            result += char(temp % 2 + '0');
        
            // Calculate the next carry
            carry = temp / 2;
        }
              
        // Reverse the string and return it
        reverse(result.begin(), result.end());
        return result;
    }
};

File: Solutions/AddBinary.py
# Leetcode #67: https://leetcode.com/problems/add-binary/
# Source: https://leetcode.com/problems/add-binary/discuss/1679423/Well-Detailed-Explaination-Java-C%2B%2B-Python-oror-Easy-for-mind-to-Accept-it

# Solution 1 | O(max(m, n)) time, O(1) space
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        # Initialize starting indices
        i, j = len(a) - 1, len(b) - 1
        carry = 0
        result = ''
        
        # Loop through both strings backwards
        while i >= 0 or j >= 0 or carry > 0:
            # Convert the letters to int and add the bits together with the carry
            if i >= 0:
                carry += int(a[i])
                i -= 1
            if j >= 0:
                carry += int(b[j])
                j -= 1
                
            # Add the sum to our string and calculate the next carry
            result += str(carry % 2)
            carry //= 2
        
        # Reverse the string b/c its backwards
        return result[::-1]
      
